pkg/apis/kudo/v1beta1/instance_admission.go

```diff
@@ -0,0 +1,239 @@
+
+ For the complete set of rules, see the corresponding test.
+*/
+
+// validateUpdate takes in the old and new (updated) instance and returns a new plan that might
+// be triggered based on the update and an error if the update is not valid. Return plan might be
+// - <nil> when there is no change to an existing scheduled plan
+// - '' empty string when an existing plan should be canceled (not implemented yet)
+// - 'newPlan' some new plan that should be triggered
+func validateUpdate(old, new *Instance, ov *OperatorVersion) (*string, error) {
+	// PREREQUISITES:
+	newPlan := new.Spec.PlanExecution.PlanName
+	oldPlan := old.Spec.PlanExecution.PlanName
+	newOvRef := new.Spec.OperatorVersion
+	oldOvRef := old.Spec.OperatorVersion
+
+	paramDiff := ParameterDiff(old.Spec.Parameters, new.Spec.Parameters)
+	paramDefs := GetParamDefinitions(paramDiff, ov)
+	triggeredPlan, err := triggeredPlan(paramDefs, ov)
+	if err != nil {
+		return nil, fmt.Errorf("failed to update Instance %s/%s: %v", old.Namespace, old.Name, err)
+	}
+
+	hasPlan := oldPlan != ""
+	isParameterUpdate := triggeredPlan != nil
+	isUpgrade := newOvRef != oldOvRef
+	isNovelPlan := !hasPlan && newPlan != ""
+	isPlanOverride := hasPlan && newPlan != "" && newPlan != oldPlan
+	isPlanCancellation := hasPlan && newPlan == ""
+
+	// Checking compatibility rules described in the table above:
+	switch {
+	case hasPlan && isParameterUpdate && *triggeredPlan != oldPlan:
+		return nil, fmt.Errorf("failed to update Instance %s/%s: plan '%s' is scheduled and an update would trigger a different plan '%s'", old.Namespace, old.Name, oldPlan, newPlan)
+	case isUpgrade && hasPlan:
+		return nil, fmt.Errorf("failed to update Instance %s/%s: upgrade to new OperatorVersion %s while a plan '%s' is scheduled is not allowed", old.Namespace, old.Name, newOvRef, oldPlan)
+	case isUpgrade && isNovelPlan:
+		return nil, fmt.Errorf("failed to update Instance %s/%s: upgrade to new OperatorVersion %s and triggering new plan '%s' is not allowed", old.Namespace, old.Name, newOvRef, newPlan)
+	case isPlanOverride:
+		return nil, fmt.Errorf("failed to update Instance %s/%s: overriding currently scheduled plan '%s' with '%s' is not supported", old.Namespace, old.Name, oldPlan, newPlan)
+	case isPlanCancellation:
+		return nil, fmt.Errorf("failed to update Instance %s/%s: cancelling currently scheduled plan '%s' is not supported", old.Namespace, old.Name, oldPlan)
+	case isParameterUpdate && isUpgrade:
+		return nil, fmt.Errorf("failed to update Instance %s/%s: upgrade to new OperatorVersion %s together with a parameter update triggering '%s' is not allowed", old.Namespace, old.Name, newOvRef, *triggeredPlan)
+	case isParameterUpdate && isNovelPlan:
+		return nil, fmt.Errorf("failed to update Instance %s/%s: triggering one plan '%s' directly and through parameter update '%s' is not allowed", old.Namespace, old.Name, oldPlan, newPlan)
+	case isParameterUpdate && isPlanOverride:
```

Comment:
this is redundant as you already have `case isPlanOverride:`

Reply:
true, because we don't support overrides right now. but, I believe, we will and then this rule will make sense 😉

Reply:
so we add it then? :) because we'll need to modify the other case then as well because right now this one will never be evaluated

Reply:
I think, plan cancelations and overrides are inevitable (but won't be a part of KEP-20 work). You can see this case as part of "future work" even if it is not evaluated yet. I can remove it, but the "next guy" might not realize it is needed (I spent quite some time thinking about all invalid update combinations)